pkg/scheduler/controller.go

```diff
@@ -0,0 +1,249 @@
+	edgeClusterClient edgeclientset.ClusterInterface
+
+	edgePlacementLister  edgev1alpha1listers.EdgePlacementClusterLister
+	edgePlacementIndexer cache.Indexer
+
+	locationLister  schedulingv1alpha1listers.LocationClusterLister
+	locationIndexer cache.Indexer
+
+	synctargetLister  workloadv1alpha1listers.SyncTargetClusterLister
+	synctargetIndexer cache.Indexer
+}
+
+func NewController(
+	context context.Context,
+	kcpClusterClient kcpclientset.ClusterInterface,
+	edgeClusterClient edgeclientset.ClusterInterface,
+	edgePlacementAccess edgev1alpha1informers.EdgePlacementClusterInformer,
+	locationAccess schedulingv1alpha1informers.LocationClusterInformer,
+	syncTargetAccess workloadv1alpha1informers.SyncTargetClusterInformer,
+) (*controller, error) {
+	context = klog.NewContext(context, klog.Background().WithValues("controller", ControllerName))
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		context: context,
+		queue:   queue,
+
+		kcpClusterClient:  kcpClusterClient,
+		edgeClusterClient: edgeClusterClient,
+
+		edgePlacementLister:  edgePlacementAccess.Lister(),
+		edgePlacementIndexer: edgePlacementAccess.Informer().GetIndexer(),
+
+		locationLister:  locationAccess.Lister(),
+		locationIndexer: locationAccess.Informer().GetIndexer(),
+
+		synctargetLister:  syncTargetAccess.Lister(),
+		synctargetIndexer: syncTargetAccess.Informer().GetIndexer(),
+	}
+
+	indexers.AddIfNotPresentOrDie(locationAccess.Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPath: indexers.IndexByLogicalClusterPath,
+	})
+
+	edgePlacementAccess.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueueEdgePlacement,
+		UpdateFunc: func(_, newObj interface{}) { c.enqueueEdgePlacement(newObj) },
+		DeleteFunc: c.enqueueEdgePlacement,
+	})
+
+	locationAccess.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueLocation,
+		UpdateFunc: func(old, obj interface{}) {
+			oldLoc := old.(*schedulingv1alpha1.Location)
+			newLoc := obj.(*schedulingv1alpha1.Location)
+			if !apiequality.Semantic.DeepEqual(oldLoc.Spec, newLoc.Spec) || !apiequality.Semantic.DeepEqual(oldLoc.Labels, newLoc.Labels) {
+				c.enqueueLocation(obj)
+			}
+		},
+		DeleteFunc: c.enqueueLocation,
+	})
+
+	syncTargetAccess.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueSyncTarget,
+		UpdateFunc: func(old, obj interface{}) {
+			oldST := old.(*workloadv1alpha1.SyncTarget)
+			newST := obj.(*workloadv1alpha1.SyncTarget)
+			if !apiequality.Semantic.DeepEqual(oldST.Spec, newST.Spec) || !apiequality.Semantic.DeepEqual(oldST.Labels, newST.Labels) {
+				c.enqueueSyncTarget((obj))
+			}
+		},
+		DeleteFunc: c.enqueueSyncTarget,
+	})
+
+	return c, nil
+}
+
+func (c *controller) enqueueEdgePlacement(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
```

Comment:
If there were an alternative that returns a struct rather than a string then the consumer would not have the expense of parsing and the complexity of handling parsing failures.  But, as far as I know, there is no such alternative.